pkg/controller/plugins.go

```diff
@@ -0,0 +1,228 @@
+	for _, plugin := range plugins {
+		name := plugin.Name()
+		if !util.IsQualifiedName(name) {
+			allErrs = append(allErrs, fmt.Errorf("controller plugin has invalid name: %#v", plugin))
+			continue
+		}
+
+		if _, found := mgr.plugins[name]; found {
+			allErrs = append(allErrs, fmt.Errorf("controller plugin %q was registered more than once", name))
+			continue
+		}
+
+		plugin.Init(host)
+		mgr.plugins[name] = &managedPlugin{plugin: plugin}
+		glog.V(1).Infof("Loaded controller plugin %q", name)
+	}
+	return errors.NewAggregate(allErrs)
+}
+
+// FindPluginByName fetches a plugin by name.  If no plugin is found, returns error.
+func (mgr *pluginMgr) FindPluginByName(name string) (Plugin, error) {
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+	if managedPlugin, ok := mgr.plugins[name]; ok {
+		return managedPlugin.plugin, nil
+	}
+	return nil, fmt.Errorf("no controller plugin found matching %s", name)
+}
+
+// RunAll attempts to create Controller instances for each plugin and run them.  This method will not fail
+// for any particular plugin.  Runtime Status is kept for all plugins.  Callers should use pluginMgr.Status
+func (mgr *pluginMgr) RunAll() {
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+	for name, managedPlugin := range mgr.plugins {
+		// do not attempt to run a plugin that had previous errors or is already running.
+		// this check makes RunAll idempotent.
+		if managedPlugin.err != nil || managedPlugin.running {
+			continue
+		}
+		managedPlugin.controller, managedPlugin.err = managedPlugin.plugin.NewController()
+		if managedPlugin.err != nil {
+			glog.V(1).Infof("Unexpected error creating controller %s: %v", name, managedPlugin.err)
+			continue
+		}
+
+		managedPlugin.err = managedPlugin.controller.Run()
+		if managedPlugin.err != nil {
+			glog.V(1).Infof("Unexpected error running controller %s: %v", name, managedPlugin.err)
+			continue
+		}
+
+		glog.V(3).Infof("Controller plugin %s is running", name)
+		managedPlugin.running = true
+	}
+}
+
+// StopAll stops all Controllers by closing their stop channels
+func (mgr *pluginMgr) StopAll() {
+	mgr.mutex.Lock()
+	defer mgr.mutex.Unlock()
+
+	for name, managedPlugin := range mgr.plugins {
+		// do not attempt to stop a plugin that isn't running or that has previous errors
+		// this checks makes StopAll idempotent
+		if managedPlugin.err != nil || !managedPlugin.running {
+			continue
+		}
+		managedPlugin.err = managedPlugin.controller.Stop()
+		if managedPlugin.err != nil {
+			glog.V(1).Infof("Unexpected error stopping controller %s: %v", name, managedPlugin.err)
+			continue
+		}
+
+		glog.V(3).Infof("Controller plugin %s is stopped", name)
+		managedPlugin.running = false
+	}
+}
+
+// Status returns the runtime status of all plugins under management
+func (mgr *pluginMgr) Status() []PluginStatus {
+	statuses := []PluginStatus{}
+	for name, managedPlugin := range mgr.plugins {
+		status := PluginStatus{
+			Name:    name,
+			Running: managedPlugin.running,
```

Comment:
This seems potentially deceptive-- you should have a way to ask the plugin what its status is, not just assume that it's running because you started it some amount of time ago.